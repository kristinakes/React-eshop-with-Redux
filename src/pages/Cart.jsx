import { useSelector } from "react-redux";
import { Link } from "react-router-dom";

import CartItem from "../components/CartItem";
import Notification from "../components/Notification";


function Cart() {
  const cartItems = useSelector((state) => state.cart.items);
  const cartTotalAmount = useSelector((state) => state.cart.totalAmount);
  const notification = useSelector((state) => state.ui.notification);

  return (
    <div>
      {notification && (
        <Notification
          title={notification.title}
          message={notification.message}
        />
      )}
      <div className="flex justify-between font-bold text-blue text-xl p-10">
        <h3>Shopping cart</h3>
        {cartItems.length !== 0 && <h3>Total: {cartTotalAmount.toFixed(2)}</h3>}
      </div>

      {cartItems.length === 0 && (
        
        <div className="text-center text-xl text-blue font-semibold rounded-md shadow-md max-w-fit p-10 m-auto">
          <p>Your shoping cart is empty. </p>
          <p>
            Go to{" "}
            <Link to="/Products" className="underline">
              Products page
            </Link>
          </p>
        </div>
      )}

      {cartItems.length > 0 && (
        <>
          {cartItems.map((item) => {
            return (
              <CartItem
                key={item.id}
                item={{
                  id: item.id,
                  title: item.title,
                  quantity: item.quantity,
                  totalPrice: item.totalPrice,
                  year: item.year,
                  ram: item["RAM"],
                  warranty: item.warranty,
                  price: item.price,
                }}
              />
            );
          })}
        </>
      )}
    </div>
  );
}
export default Cart;
