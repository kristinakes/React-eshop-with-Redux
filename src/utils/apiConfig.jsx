export const API_PRODUCTS_URL = "http://localhost:8000/products/";
export const API_CART_URL = "http://localhost:8000/cart/";