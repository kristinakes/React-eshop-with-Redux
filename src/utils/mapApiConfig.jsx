//Google maps API key to be used in contacts page. Key should be protected. Update accordingly.
///More info https://developers.google.com/maps/documentation/javascript/get-api-key
export const GOOGLE_API_KEY = "Add your API Key here";