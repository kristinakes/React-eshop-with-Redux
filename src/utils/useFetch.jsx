import { useState, useEffect } from "react";

function useFetch(url, options) {
  const [data, setData] = useState(null);
  const [isPending, setIsPending] = useState(true);
  const [error, setError] = useState(null);

  useEffect(()=> {
    setTimeout(()=> {
        fetch(url, options).then(result => {
            if (!result.ok) {
                throw Error('Error fetching data');
            }
            return result.json();
        }).then(data => {
            setData(data);
            setIsPending(false);
            setError(null);
        }).catch(err => {
            setIsPending(false);
            setError(err.message)
        });
    }, 1000);
  }, [url, options]);

  return {data, isPending, error};
}
export default useFetch;
