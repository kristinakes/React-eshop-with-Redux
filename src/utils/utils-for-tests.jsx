import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import cartSlice from "../store/cartSlice";
import uiSlice from "../store/uiSlice";
import productSlice from "../store/productSlice"

export function renderWithProviders(
  ui,
  {
    preloadedState = {},

    store = configureStore({
      reducer: {
        products: productSlice.reducer,
        cart: cartSlice.reducer,
        ui: uiSlice.reducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
