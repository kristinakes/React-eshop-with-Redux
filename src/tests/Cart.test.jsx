import { renderWithProviders } from "../utils/utils-for-tests";
import { MemoryRouter } from "react-router-dom";
import { screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Cart from "../pages/Cart";

const cartState = {
  id: 1,
  items: [
    {
      id: 1,
      title: "laptop",
      price: 10,
      year: 2020,
      RAM: "5",
      warranty: "2 years",
      totalPrice: 20,
      quantity: 2,
    },
  ],
  totalQuantity: 2,
  totalAmount: 20,
};

describe("Cart page", () => {
  test("should render page correctly when there is no items in cart", () => {
    renderWithProviders(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(screen.getByText(/cart is empty/i)).toBeInTheDocument();
    const link = screen.getByRole("link", { name: /products page/i });
    expect(link).toHaveAttribute("href", "/Products");
  });

  // test("should render item in cart if there are items in store", () => {
  //   renderWithProviders(
  //     <MemoryRouter>
  //       <Cart />
  //     </MemoryRouter>,
  //     {
  //       preloadedState: { cart: cartState },
  //     }
  //   );

  //   //Assert if total cart price rendered correctly
  //    const totalPrice = screen.getByRole('heading', {name: /total/i});
  //    expect(totalPrice).toHaveTextContent(cartState.totalAmount);
  //     //Assert if product image is rendered
  //   expect(screen.getByRole('img')).toBeInTheDocument();
  // });

  
  //Snapshot test for cart page
  test('should match snapshot if item exists in cart store', () => { 
    const view = renderWithProviders(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      {
        preloadedState: { cart: cartState },
      }
    );

    expect(view.baseElement).toMatchSnapshot();
  })

  test('should increase/decrease product quantity in cart', async() => { 
    const {store} = renderWithProviders(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      {
        preloadedState: { cart: cartState },
      }
    );

    const buttons = screen.getAllByRole('button');
    await fireEvent.click(buttons[1]);  

    expect(store.getState().cart.totalQuantity).toBe(3);
    expect(store.getState().cart.totalAmount).toBe(30);

    await fireEvent.click(buttons[0]);  
    expect(store.getState().cart.totalQuantity).toBe(2);

  })
});
