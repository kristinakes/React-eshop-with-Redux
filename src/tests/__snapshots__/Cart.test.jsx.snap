// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart page should match snapshot if item exists in cart store 1`] = `
<body>
  <div>
    <div>
      <div
        class="flex justify-between font-bold text-blue text-xl p-10"
      >
        <h3>
          Shopping cart
        </h3>
        <h3>
          Total: 
          20.00
        </h3>
      </div>
      <div
        class="grid grid-cols-2 gap-4 min-w-fit justify-center content-center sm:grid-cols-4 sm:gap-2 shadow-md m-5 p-3"
      >
        <div
          class="size-48 m-auto"
        >
          <a
            href="/Product/1"
          >
            <img
              alt="laptop"
              class="object-fill scale-75 z-0"
              src="/assets/product_images/laptop.png"
            />
          </a>
        </div>
        <div>
          <a
            href="/Product/1"
          >
            <h3
              class="font-bold text-xl text-blue my-5"
            >
              laptop
            </h3>
          </a>
          <div
            class="font-semibold flex flex-wrap gap-3 md:flex-col md:gap-1"
          >
            <p>
              Year: 
              2020
            </p>
            <p>
              RAM memory: 
              5
            </p>
            <p>
              Warranty: 
              2 years
            </p>
            <p>
              Price: 
              10
            </p>
          </div>
        </div>
        <div
          class="flex justify-center items-center"
        >
          <button
            class="btn-blue"
            id="remove-item-btn"
          >
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <p
            class="font-semibold text-4xl mx-5"
          >
            2
          </p>
          <button
            class="btn-transparent"
            id="add-item-btn"
          >
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 4.5v15m7.5-7.5h-15"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </div>
        <div
          class="font-bold text-xl m-auto"
        >
          <p>
            Total: 
            20.00
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
